Objects/funcobject.c

```diff
@@ -766,7 +795,14 @@ func_clear(PyFunctionObject *op)
 static void
 func_dealloc(PyFunctionObject *op)
 {
+    assert(Py_REFCNT(op) == 0);
+    Py_SET_REFCNT(op, 1);
     handle_func_event(PyFunction_EVENT_DESTROY, op, NULL);
+    if (Py_REFCNT(op) > 1) {
+        Py_SET_REFCNT(op, Py_REFCNT(op) - 1);
+        return;
+    }
+    Py_SET_REFCNT(op, 0);
```

Comment:
Is this `Py_SET_REFCNT` necessary? IIUC this is redundant here otherwise it isn't obvious to me and a comment would be nice.

Reply:
In the normal case, assuming the watcher has not taken any new reference to the object, the refcount will be `1` at this point (because we set it to 1 above.) We should set it to `0` before continuing with deallocation, because objects being deallocated should have a refcount of zero.

Reply:
The watcher *must* take a reference to any object it is watching, otherwise the refcount is incorrect and we crash and burn.

The GC, or an optimizer, can legitimately assume that any object with a reference count of zero and no weakrefs is to be freed and re-use the memory. That's not going to end well if the watcher still thinks that memory is still a dictionary.

Reply:
@markshannon I think your comment is more relevant to the discussion below (about why to have dealloc watchers at all) than it is to this thread, so I'll reply below.

The answer to @kumaraditya303's question here is that this `Py_SET_REFCNT(op, 0)` is the decref that has to be paired with the incref `Py_SET_REFCNT(op, 1);` that occurs on line 799. We use explicit `Py_SET_REFCNT` instead of `Py_INCREF` and `Py_DECREF` because `Py_DECREF` would cause recursive deallocation, and if we use `Py_SET_REFCNT` for one we have to use it for the other also, otherwise the total refcount tracking will be off.